File: legacy/SemanticScript/bench/algorithms/collatz.py
"""collatz.py - sum of Collatz stopping times for every start in 1..N.

Mirrors collatz.c's algorithm and checksum. Uses bitwise parity/halving
(`value & 1`, `value >> 1`), which is correct for Python's arbitrary-precision
ints and faster than `% 2` / `// 2`. (The JavaScript port cannot do this: its
bitwise operators are 32-bit and Collatz peaks exceed 2^31.) Timed with
time.perf_counter().
"""

import time


def main():
    n = 1000000
    total_steps = 0

    start = time.perf_counter()
    for k in range(1, n + 1):
        value = k
        steps = 0
        while value != 1:
            if value & 1:
                value = 3 * value + 1
            else:
                value >>= 1
            steps += 1
        total_steps += steps
    end = time.perf_counter()

    elapsed_seconds = end - start
    print(f"checksum={total_steps} elapsedSeconds={elapsed_seconds:.6f}")


if __name__ == "__main__":
    main()

File: legacy/SemanticScript/bench/algorithms/collatz.js
// collatz.js - sum of Collatz stopping times for every start in 1..N.
// Mirrors collatz.c: same N, same checksum, same output contract.
//
// Intermediate Collatz values for N < 10^6 reach ~10^11, which exceeds 2^31,
// so this MUST use arithmetic (% 2, / 2), not bitwise operators: JS bitwise
// ops coerce to 32-bit and would corrupt the values. Plain JS numbers are
// exact here because every value stays below 2^53. The hot loop is wrapped in
// main() so V8 JIT-optimizes it.

function main() {
  const n = 1000000;
  let totalSteps = 0;

  const start = performance.now();
  for (let k = 1; k <= n; k++) {
    let value = k;
    let steps = 0;
    while (value !== 1) {
      if (value % 2 === 0) {
        value = value / 2;
      } else {
        value = 3 * value + 1;
      }
      steps++;
    }
    totalSteps += steps;
  }
  const end = performance.now();

  const elapsedSeconds = (end - start) / 1000;
  console.log(`checksum=${totalSteps} elapsedSeconds=${elapsedSeconds.toFixed(6)}`);
}

main();
